### PR TITLE
Add native histograms to mimirpb

### PR DESCRIPTION
**Pull request by codesome**
#### What this PR does

Syncs changes done to `mimirpb` package in `sparsehistogram` branch and also syncs other packages that break with this change.

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/mimirpb/compat.go [hunk 1]
```diff
@@ -593,20 +665,32 @@ loop:
 	return numLabels, true
 }
 
-// Generics
 type GenericSamplePair interface {
```
Comment (krajorama) [inline pkg/mimirpb/compat.go:668]: I don't think this is used yet and also needs #4191 , so I'd just remove the whole GenericSamplePair code starting from line 668(new) for now.

Comment (codesome) [reply pkg/mimirpb/compat.go:668]: `GenericSamplePair` is already in the `main` branch, so I don't think removing and adding back is worth it. #4191 can be directly done in main branch now, isn't it?

File: pkg/mimirpb/compat.go [hunk 2]
```diff
@@ -299,6 +310,67 @@ func fromSpansToSpansProto(s []histogram.Span) []BucketSpan {
...
+		samples = append(samples, FloatHistogramPair{
+			TimestampMs: point.T,
+			Histogram:   *FloatHistogramFromPrometheusModel(point.H),
+		})
+	}
+	return samples
+}
+
+// FromFloatHistogramToPromHistogram converts histogram.FloatHistogram to model.SampleHistogram.
+func FromFloatHistogramToPromHistogram(h *histogram.FloatHistogram) *model.SampleHistogram {
```
Comment (krajorama) [inline pkg/mimirpb/compat.go:345]: I don't think we're testing this directly. Although it's used in generating a frontend reply, so integration tests will cover it. We should add a small smoke test though.

**Review by krajorama — APPROVED** [review]
Couple of minor issues, but LGTM

File: pkg/mimirpb/compat.go (hunk 1 above)
Comment (krajorama) [reply pkg/mimirpb/compat.go:668]: true, let's just keep it in mind

**Pull request merged**